File: hash_table/maximum_subarray_sum.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def maximum_subarray_sum(nums, k)
  hash_table = {}
  max = nil

  prefix_sums = [0]
  nums.each_with_index do |num, index|
    prefix_sums << prefix_sums.last + num
  end

  nums.each_with_index do |num, index|
    guesses = [hash_table[num + k], hash_table[num - k]].compact.flatten

    guesses.each do |guess|
      # prefix sum of current index minus the guess prefix sum (start of subarray)
      new_max = prefix_sums[index + 1] - guess
      max = new_max if max.nil? || new_max > max
    end

    if hash_table[num]
      # only store the smallest index prefix sum, as we will decrease it later
      hash_table[num] = [prefix_sums[index], hash_table[num]].min
    else
      hash_table[num] = prefix_sums[index]
    end
  end
  max || 0
end

File: hash_table/maximum_subarray_sum.py
class Solution(object):
  def maximumSubarraySum(self, nums, k):
    hash_table = {}
    max_sum = None

    prefix_sums = [0]
    for num in nums:
      prefix_sums.append(prefix_sums[-1] + num)

    for index, num in enumerate(nums):
      if hash_table.get(num + k) is not None:
        new_max = prefix_sums[index + 1] - hash_table.get(num + k)
        if max_sum is None or new_max > max_sum:
          max_sum = new_max

      if hash_table.get(num - k) is not None:
        new_max = prefix_sums[index + 1] - hash_table.get(num - k)
        if max_sum is None or new_max > max_sum:
          max_sum = new_max

      if hash_table.get(num) is not None:
        # only store the smallest index prefix sum, as we will decrease it later
        hash_table[num] = min(prefix_sums[index], hash_table[num])
      else:
        hash_table[num] = prefix_sums[index]

    return max_sum if max_sum is not None else 0
